File: src/parser/lexer.ts
import type { SourceSpan } from "../model.js";

export type TokenKind =
  | "number"
  | "string"
  | "true"
  | "false"
  | "identifier"
  | "return"
  | "let"
  | "const"
  | "if"
  | "else"
  | "this"
  | "pipeOp"
  | "dot"
  | "comma"
  | "colon"
  | "semicolon"
  | "lparen"
  | "rparen"
  | "lbrace"
  | "rbrace"
  | "lbracket"
  | "rbracket"
  | "plus"
  | "minus"
  | "star"
  | "slash"
  | "percent"
  | "eq"
  | "eqeq"
  | "neq"
  | "lt"
  | "gt"
  | "lte"
  | "gte"
  | "and"
  | "or"
  | "bang"
  | "pipe"
  | "arrow"
  | "fatArrow"
  | "error"
  | "question"
  | "eof";

export type Token = {
  kind: TokenKind;
  value: string;
  span: SourceSpan;
};

const KEYWORDS: Record<string, TokenKind> = {
  return: "return",
  let: "let",
  const: "const",
  if: "if",
  else: "else",
  this: "this",
  true: "true",
  false: "false",
};

const MULTI_CHAR_OPS: Record<string, Record<string, TokenKind>> = {
  "=": { "=": "eqeq" },
  "!": { "=": "neq" },
  "<": { "=": "lte" },
  ">": { "=": "gte" },
  "&": { "&": "and" },
  "|": { "|": "or", ">": "pipeOp" },
  "-": { ">": "fatArrow" },
};

function isDigit(ch: string): boolean {
  return ch >= "0" && ch <= "9";
}

function isIdentifierStart(ch: string): boolean {
  return (ch >= "a" && ch <= "z") || (ch >= "A" && ch <= "Z") || ch === "_";
}

function isIdentifierPart(ch: string): boolean {
  return isIdentifierStart(ch) || isDigit(ch);
}

export function tokenize(source: string, file: string): Token[] {
  const tokens: Token[] = [];
  let pos = 0;
  let line = 1;
  let column = 1;

  function pushToken(
    kind: TokenKind,
    value: string,
    sl: number,
    sc: number,
    el: number,
    ec: number,
  ): void {
    tokens.push({
      kind,
      value,
      span: { file, startLine: sl, startColumn: sc, endLine: el, endColumn: ec },
    });
  }

  function peek(offset: number): string | undefined {
    const idx = pos + offset;
    return idx < source.length ? source[idx] : undefined;
  }

  while (pos < source.length) {
    const startLine = line;
    const startColumn = column;
    const ch = source[pos]!;

    // ── Whitespace ──────────────────────────────────────
    if (ch === " " || ch === "\t" || ch === "\r") {
      pos++;
      column++;
      continue;
    }
    if (ch === "\n") {
      pos++;
      line++;
      column = 1;
      continue;
    }

    // ── Numbers (integer and decimal) ───────────────────
    if (isDigit(ch)) {
      let value = "";
      while (pos < source.length && isDigit(source[pos]!)) {
        value += source[pos];
        pos++;
        column++;
      }
      if (
        peek(0) === "." &&
        peek(1) !== undefined &&
        isDigit(peek(1)!)
      ) {
        value += ".";
        pos++;
        column++;
        while (pos < source.length && isDigit(source[pos]!)) {
          value += source[pos];
          pos++;
          column++;
        }
      }
      pushToken("number", value, startLine, startColumn, line, column);
      continue;
    }

    // ── Strings (double-quoted) ─────────────────────────
    if (ch === '"') {
      let value = '"';
      pos++; // skip opening "
      column++;
      while (pos < source.length && source[pos] !== '"') {
        if (source[pos] === "\n") {
          line++;
          column = 1;
        } else {
          column++;
        }
        value += source[pos];
        pos++;
      }
      if (pos < source.length && source[pos] === '"') {
        value += '"';
        pos++; // skip closing "
        column++;
        pushToken("string", value, startLine, startColumn, line, column);
      } else {
        pushToken("error", "Unterminated string literal", startLine, startColumn, line, column);
      }
      continue;
    }

    // ── Identifiers / Keywords ──────────────────────────
    if (isIdentifierStart(ch)) {
      let value = "";
      while (pos < source.length && isIdentifierPart(source[pos]!)) {
        value += source[pos];
        pos++;
        column++;
      }
      const kind: TokenKind = KEYWORDS[value] ?? "identifier";
      pushToken(kind, value, startLine, startColumn, line, column);
      continue;
    }

    // ── Multi-character operators (checked before single) ──
    const next = peek(1);
    if (next !== undefined) {
      const multiKind: TokenKind | undefined = MULTI_CHAR_OPS[ch]?.[next];
      if (multiKind !== undefined) {
        pushToken(multiKind, ch + next, startLine, startColumn, line, column + 2);
        pos += 2;
        column += 2;
        continue;
      }
    }

    // ── Single-character operators and delimiters ────────
    const singleKind: TokenKind | undefined = ((): TokenKind | undefined => {
      switch (ch) {
        case "+": return "plus";
        case "-": return "minus";
        case "*": return "star";
        case "/": return "slash";
        case "%": return "percent";
        case "=": return "eq";
        case "!": return "bang";
        case "<": return "lt";
        case ">": return "gt";
        case "|": return "pipe";
        case ".": return "dot";
        case ",": return "comma";
        case ":": return "colon";
        case ";": return "semicolon";
        case "(": return "lparen";
        case ")": return "rparen";
        case "{": return "lbrace";
        case "}": return "rbrace";
        case "[": return "lbracket";
        case "]": return "rbracket";
        case "?": return "question";
        default: return undefined;
      }
    })();

    if (singleKind !== undefined) {
      pushToken(singleKind, ch, startLine, startColumn, line, column + 1);
      pos++;
      column++;
      continue;
    }

    // ── Unknown character ────────────────────────────────
    pushToken("error", `Unexpected character: '${ch}'`, startLine, startColumn, line, column + 1);
    pos++;
    column++;
  }

  // EOF sentinel
  pushToken("eof", "", line, column, line, column);

  return tokens;
}

File: tangle-cli/src/parser/lexer.rs
use crate::model::{SourceSpan, TangleDiagnostic};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TokenKind {
    Number, String, True, False, Identifier,
    Return, Let, Const, If, Else, This,
    PipeOp, Dot, Comma, Colon, Semicolon,
    LParen, RParen, LBrace, RBrace, LBracket, RBracket,
    Plus, Minus, Star, Slash, Percent,
    Eq, EqEq, Neq, Lt, Gt, Lte, Gte,
    And, Or, Bang, Pipe, Arrow, FatArrow,
    Question, ErrorKw, Eof,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Token {
    pub kind: TokenKind,
    pub lexeme: String,
    pub span: SourceSpan,
}

pub struct Lexer<'a> {
    #[allow(dead_code)]
    source: &'a str,
    file: &'a str,
    chars: Vec<char>,
    pos: usize,
    line: usize,
    col: usize,
    diagnostics: Vec<TangleDiagnostic>,
}

impl<'a> Lexer<'a> {
    pub fn new(source: &'a str, file: &'a str) -> Self {
        Lexer { source, file, chars: source.chars().collect(), pos: 0, line: 1, col: 1, diagnostics: vec![] }
    }

    pub fn diagnostics(&self) -> &[TangleDiagnostic] { &self.diagnostics }

    fn current(&self) -> Option<char> { self.chars.get(self.pos).copied() }
    fn peek(&self, offset: usize) -> Option<char> { self.chars.get(self.pos + offset).copied() }

    fn advance(&mut self) -> Option<char> {
        let c = self.chars.get(self.pos).copied();
        if let Some(ch) = c {
            self.pos += 1;
            if ch == '\n' { self.line += 1; self.col = 1; }
            else { self.col += 1; }
        }
        c
    }

    fn make_span(&self, sl: usize, sc: usize, el: usize, ec: usize) -> SourceSpan {
        SourceSpan { file: self.file.to_string(), start_line: sl, start_column: sc, end_line: el, end_column: ec }
    }

    pub fn tokenize(&mut self) -> Vec<Token> {
        let mut tokens = vec![];
        loop {
            self.skip_whitespace();
            let sl = self.line; let sc = self.col;
            let c = match self.current() { Some(c) => c, None => break };
            let token = match c {
                '0'..='9' => self.read_number(sl, sc),
                '"' => self.read_string(sl, sc),
                '|' if self.peek(1) == Some('>') => { self.advance(); self.advance(); self.make_tok(TokenKind::PipeOp, "|>", sl, sc) }
                '|' if self.peek(1) == Some('|') => { self.advance(); self.advance(); self.make_tok(TokenKind::Or, "||", sl, sc) }
                '|' => { self.advance(); self.make_tok(TokenKind::Pipe, "|", sl, sc) }
                '=' if self.peek(1) == Some('=') => { self.advance(); self.advance(); self.make_tok(TokenKind::EqEq, "==", sl, sc) }
                '=' if self.peek(1) == Some('>') => { self.advance(); self.advance(); self.make_tok(TokenKind::FatArrow, "=>", sl, sc) }
                '=' => { self.advance(); self.make_tok(TokenKind::Eq, "=", sl, sc) }
                '!' if self.peek(1) == Some('=') => { self.advance(); self.advance(); self.make_tok(TokenKind::Neq, "!=", sl, sc) }
                '!' => { self.advance(); self.make_tok(TokenKind::Bang, "!", sl, sc) }
                '<' if self.peek(1) == Some('=') => { self.advance(); self.advance(); self.make_tok(TokenKind::Lte, "<=", sl, sc) }
                '<' => { self.advance(); self.make_tok(TokenKind::Lt, "<", sl, sc) }
                '>' if self.peek(1) == Some('=') => { self.advance(); self.advance(); self.make_tok(TokenKind::Gte, ">=", sl, sc) }
                '>' => { self.advance(); self.make_tok(TokenKind::Gt, ">", sl, sc) }
                '&' if self.peek(1) == Some('&') => { self.advance(); self.advance(); self.make_tok(TokenKind::And, "&&", sl, sc) }
                '-' if self.peek(1) == Some('>') => { self.advance(); self.advance(); self.make_tok(TokenKind::Arrow, "->", sl, sc) }
                '-' => { self.advance(); self.make_tok(TokenKind::Minus, "-", sl, sc) }
                '.' => { self.advance(); self.make_tok(TokenKind::Dot, ".", sl, sc) }
                ',' => { self.advance(); self.make_tok(TokenKind::Comma, ",", sl, sc) }
                ':' => { self.advance(); self.make_tok(TokenKind::Colon, ":", sl, sc) }
                ';' => { self.advance(); self.make_tok(TokenKind::Semicolon, ";", sl, sc) }
                '(' => { self.advance(); self.make_tok(TokenKind::LParen, "(", sl, sc) }
                ')' => { self.advance(); self.make_tok(TokenKind::RParen, ")", sl, sc) }
                '{' => { self.advance(); self.make_tok(TokenKind::LBrace, "{", sl, sc) }
                '}' => { self.advance(); self.make_tok(TokenKind::RBrace, "}", sl, sc) }
                '[' => { self.advance(); self.make_tok(TokenKind::LBracket, "[", sl, sc) }
                ']' => { self.advance(); self.make_tok(TokenKind::RBracket, "]", sl, sc) }
                '+' => { self.advance(); self.make_tok(TokenKind::Plus, "+", sl, sc) }
                '*' => { self.advance(); self.make_tok(TokenKind::Star, "*", sl, sc) }
                '/' => { self.advance(); self.make_tok(TokenKind::Slash, "/", sl, sc) }
                '%' => { self.advance(); self.make_tok(TokenKind::Percent, "%", sl, sc) }
                '?' => { self.advance(); self.make_tok(TokenKind::Question, "?", sl, sc) }
                c if c.is_alphabetic() || c == '_' => self.read_identifier_or_keyword(sl, sc),
                _ => {
                    self.advance();
                    self.diagnostics.push(TangleDiagnostic { code: "TANGLE_LEXER_ERROR".into(), message: format!("Unknown character: '{}'", c), span: self.make_span(sl, sc, self.line, self.col) });
                    Token { kind: TokenKind::Eof, lexeme: c.to_string(), span: self.make_span(sl, sc, self.line, self.col) }
                }
            };
            tokens.push(token);
        }
        tokens.push(Token { kind: TokenKind::Eof, lexeme: String::new(), span: self.make_span(self.line, self.col, self.line, self.col) });
        tokens
    }

    fn skip_whitespace(&mut self) {
        while let Some(c) = self.current() {
            if c == ' ' || c == '\t' || c == '\r' || c == '\n' { self.advance(); } else { break; }
        }
    }

    fn read_number(&mut self, sl: usize, sc: usize) -> Token {
        let mut lexeme = String::new();
        while let Some(c) = self.current() {
            if c.is_ascii_digit() || c == '.' { lexeme.push(c); self.advance(); } else { break; }
        }
        Token { kind: TokenKind::Number, lexeme, span: self.make_span(sl, sc, self.line, self.col) }
    }

    fn read_string(&mut self, sl: usize, sc: usize) -> Token {
        self.advance();
        let mut lexeme = String::new();
        while let Some(c) = self.current() {
            if c == '"' { self.advance(); return Token { kind: TokenKind::String, lexeme, span: self.make_span(sl, sc, self.line, self.col) }; }
            lexeme.push(c); self.advance();
        }
        self.diagnostics.push(TangleDiagnostic { code: "TANGLE_UNTERMINATED_STRING".into(), message: "Unterminated string literal".into(), span: self.make_span(sl, sc, self.line, self.col) });
        Token { kind: TokenKind::String, lexeme, span: self.make_span(sl, sc, self.line, self.col) }
    }

    fn read_identifier_or_keyword(&mut self, sl: usize, sc: usize) -> Token {
        let mut lexeme = String::new();
        while let Some(c) = self.current() {
            if c.is_alphanumeric() || c == '_' { lexeme.push(c); self.advance(); } else { break; }
        }
        let kind = match lexeme.as_str() {
            "return" => TokenKind::Return, "let" => TokenKind::Let, "const" => TokenKind::Const,
            "if" => TokenKind::If, "else" => TokenKind::Else, "this" => TokenKind::This,
            "true" => TokenKind::True, "false" => TokenKind::False,
            "match" => TokenKind::ErrorKw, "panic" => TokenKind::ErrorKw,
            _ => TokenKind::Identifier,
        };
        Token { kind, lexeme, span: self.make_span(sl, sc, self.line, self.col) }
    }

    fn make_tok(&self, kind: TokenKind, lexeme: &str, sl: usize, sc: usize) -> Token {
        Token { kind, lexeme: lexeme.to_string(), span: self.make_span(sl, sc, self.line, self.col) }
    }
}

pub fn tokenize(source: &str, file: &str) -> (Vec<Token>, Vec<TangleDiagnostic>) {
    let mut lexer = Lexer::new(source, file);
    let tokens = lexer.tokenize();
    let diagnostics = lexer.diagnostics;
    (tokens, diagnostics.to_vec())
}
